packages/embark-ui/src/containers/DeploymentContainer.js

```diff
@@ -1,35 +1,88 @@
 import React, {Component} from 'react';
 import {connect} from 'react-redux';
 import PropTypes from 'prop-types';
-import {
-  contracts as contractsAction,
-  web3Deploy as web3DeployAction,
-  web3EstimateGas as web3EstimateGasAction,
-  updateDeploymentPipeline} from "../actions";
-
+import {contracts as contractsAction,
+        listenToContracts,
+        stopContracts,
+        web3Deploy as web3DeployAction,
+        web3EstimateGas as web3EstimateGasAction,
+        updateDeploymentPipeline} from "../actions";
 import ContractsDeployment from '../components/ContractsDeployment';
 import DataWrapper from "../components/DataWrapper";
-import PageHead from '../components/PageHead';
 import {getContracts, getDeploymentPipeline, getWeb3, getWeb3GasEstimates, getWeb3Deployments} from "../reducers/selectors";
+import PageHead from '../components/PageHead';
+
+const MAX_CONTRACTS = 10;
 
 class DeploymentContainer extends Component {
+  constructor(props) {
+    super(props);
+
+    this.numContractsToDisplay = this.props.numContractsToDisplay || MAX_CONTRACTS;
+    this.state = {currentPage: 1};
+  }
+
   componentDidMount() {
     this.props.fetchContracts();
+    this.props.listenToContracts();
+  }
+
+  componentWillUnmount() {
+    this.props.stopContracts();
+  }
+
+  get numberOfContracts() {
+    return this.props.contracts
+      .filter(contract => (contract.code || contract.deploy) && !contract.silent)
+      .length;
+  }
+
+  getNumberOfPages() {
+    return Math.ceil(this.numberOfContracts / this.numContractsToDisplay);
+  }
+
+  changePage(newPage) {
+    this.setState({currentPage: newPage});
+    this.props.fetchContracts();
+  }
+
+  getCurrentContracts() {
+    let offset = 0;
+    return this.props.contracts
+      .filter((contract, arrIndex) => {
+        if (!(contract.code || contract.deploy) || contract.silent) {
+          offset++;
+          return false
+        };
+        const index = (
+          (arrIndex + 1 - offset) -
+            (this.numContractsToDisplay * (this.state.currentPage - 1))
+        );
+        return index <= this.numContractsToDisplay && index > 0;
+      });
```

Comment:
There is a lot of duplication of functions between the two containers. Do you think there would be a way to use inheritance or maybe more simply put the shared functions in a utils file?

Reply:
You're right. We've got five containers that have similar functions. However, they're all slightly different in terms of what props they refer to and/or how the calculations are made.  Refactoring would need to account for those differences somehow and I wonder if it that would trade complexity for DRY-ness.

Reply:
Ah you're right. Having to manage the props or passing them in the functions could be a pain. 
Anyway, I'll let you decide if you want to do it or not

Reply:
There are some opportunities for reducing duplication and also introducing some optimizations. See, for example, changes re: `resetNums()`. I'll make the changes in another PR, for all of the relevant containers.